python/plugins/processing/algs/qgis/SplitWithLines.py

```diff
@@ -0,0 +1,175 @@
+# -*- coding: utf-8 -*-
+
+"""
+***************************************************************************
+    SplitWithLines.py
+    ---------------------
+    Date                 : November 2014
+    Revised              : September 2016
+    Copyright            : (C) 2014 by Bernhard Ströbl
+    Email                : bernhard dot stroebl at jena dot de
+***************************************************************************
+*                                                                         *
+*   This program is free software; you can redistribute it and/or modify  *
+*   it under the terms of the GNU General Public License as published by  *
+*   the Free Software Foundation; either version 2 of the License, or     *
+*   (at your option) any later version.                                   *
+*                                                                         *
+***************************************************************************
+"""
+
+__author__ = 'Bernhard Ströbl'
+__date__ = 'November 2014'
+__copyright__ = '(C) 2014, Bernhard Ströbl'
+
+# This will get replaced with a git SHA1 when you do a git archive
+
+__revision__ = '$Format:%H$'
+
+from qgis.core import Qgis, QgsFeatureRequest, QgsFeature, QgsGeometry, QgsWkbTypes
+from processing.core.GeoAlgorithm import GeoAlgorithm
+from processing.core.parameters import ParameterVector
+from processing.core.outputs import OutputVector
+from processing.core.ProcessingLog import ProcessingLog
+from processing.tools import dataobjects
+from processing.tools import vector
+
+
+class SplitWithLines(GeoAlgorithm):
+
+    INPUT_A = 'INPUT_A'
+    INPUT_B = 'INPUT_B'
+
+    OUTPUT = 'OUTPUT'
+
+    def defineCharacteristics(self):
+        self.name, self.i18n_name = self.trAlgorithm('Split with lines')
+        self.group, self.i18n_group = self.trAlgorithm('Vector overlay tools')
+        self.addParameter(ParameterVector(self.INPUT_A,
+                                          self.tr('Input layer, single geometries only'), [dataobjects.TYPE_VECTOR_POLYGON,
+                                          dataobjects.TYPE_VECTOR_LINE]))
+        self.addParameter(ParameterVector(self.INPUT_B,
+                                          self.tr('Split layer'), [dataobjects.TYPE_VECTOR_LINE]))
+
+        self.addOutput(OutputVector(self.OUTPUT, self.tr('Splitted')))
+
+    def processAlgorithm(self, progress):
+        layerA = dataobjects.getObjectFromUri(self.getParameterValue(self.INPUT_A))
+        splitLayer = dataobjects.getObjectFromUri(self.getParameterValue(self.INPUT_B))
+
+        sameLayer = self.getParameterValue(self.INPUT_A) == self.getParameterValue(self.INPUT_B)
+        fieldList = layerA.fields()
+
+        writer = self.getOutputFromName(self.OUTPUT).getVectorWriter(fieldList,
+                                                                     layerA.wkbType(), layerA.crs())
+
+        spatialIndex = vector.spatialindex(splitLayer)
```

Comment:
Thanks! The inbuilt processing methods are fine for simple use cases (eg looping once through features and applying an operation to each), but to really optimise complex algs you need to use the  qgis api itself.